src/qibo/backends/numpy.py

```diff
@@ -30,6 +30,8 @@ def __init__(self):
             np.complex128,
         )
 
+        self.clifford_operations = _clifford_operations
```

Comment:
This should be in the `CliffordBackend`, NumPy doesn't have to know nothing about Clifford.

Reply:
I mean each backend (engine) has it's own definition of the `clifford_operations`. When you create the `CliffordBackend` you also decide which engine to use and then it's up to engine to define how to perform the clifford operations.

Reply:
There's a problem in the design of `CliffordBackend`: I didn't notice that `CliffordBackend` is both subclassing `NumpyBackend` and setting a backend as an engine.

So, in the case of CuPy, you would have a `CliffordBackend` with an engine being `CupyBackend`, and `clifford_operations_gpu` as operations, but still subclassing `NumpyBackend` directly.

I believe it is too much tangled, but possibly error-prone.

Reply:
However, here the main issue is isolation: having Clifford related functions where Clifford is potentially not involved is anything but modular.

E.g. the call to `_clifford_pre_execution_reshape` (comment below) is 
https://github.com/qiboteam/qibo/blob/9d15645e6fe3a3629bedd0ab3d95541f71d75e9a/src/qibo/backends/clifford.py#L158
that is delegating to
https://github.com/qiboteam/qibo/blob/9d15645e6fe3a3629bedd0ab3d95541f71d75e9a/src/qibo/backends/clifford.py#L97

But in principle you could directly invoke a function from the suitable operations module, or something like.

Reply:
In this case in particular, you're setting the operations in the engine, and the engine in `CliffordBackend`.

I know the engine is useful, since you could use to cast and things like that. Though, you're getting the same through subclassing.
But in case you still want to have the engine around as you set the engine you could also set the operations, and you could just expose a variable with a tuple `(engine, operations_module)`, that allows to avoid engine modifications and still set both in a single shot.
Or instead revert the dependency, and put an `ENGINE` constant in the operations modules, intended as default engine in case those operations are chosen.

Reply:
Although I agree that inheriting from the numpy backend and having a different engine inside is not the cleanest solution. I am not sure I agree that having `clifford` related functions in non clifford backends is a problem. I mean in the end the `CliffordSimulator` is just an alternative to state vector simulation, and you still have all the ingredients to perform density matrix and state vector evolution in the backends. 
Anyway, in general I think that this is a bit out of the scope of this PR, whose sole aim was accommodating for the updates needed by the  PR in `qibojit`.

Reply:
But in main it is already like that. The `CliffordBackend` inherits from the `NumpyBackend` this PR is not changing anything in that sense.

Reply:
I believe that this is a similar problem to the one we encountered in the cloud backends PR. Indeed, those two inherit from the `NumpyBackend` already, but ideally they wouldn't need to. Therefore, I was thinking, that this might be a good moment to start thinking about having a more comprehensive abstract backend which will take care of the more upper level functions, like for instance: samples generation, probabilities and frequencies calculation.

Reply:
Some of those backend functions could be moved to `quantum_info` as suggested [here](https://github.com/qiboteam/qibo/issues/1168)

Reply:
> Some of those backend functions could be moved to `quantum_info` as suggested [here](https://github.com/qiboteam/qibo/issues/1168)

@renatomello, the only problem of moving functions out of the backends is that you will then have a single implementation (having multiple implementations is the raison d'être of the backends' mechanism).
However, if we have functions that are not performance-intensive, and which do not require optimized implementations, they certainly do not belong to the backends in the first place.

Reply:
Then I am getting a bit lost. So you are saying that the problem is the presence of the various clifford related functions in the backends? Like for example the clifford reshape and the `clifford_operations` attribute. But then, I did not get what is your proposed solution.

Reply:
The problem is the presence of these functions (and invocations) in the `NumpyBackend`.

The solution is to make the `CliffordBackend` directly aware of the `clifford_` functions to be invoked, not passing through the `engine`.
The reason you're passing through the engine is that you have multiple implementations of these functions, and one single `CliffordBackend`. But if the Clifford backend is aware of the Clifford operations module separately from the backend, then you don't need any longer to pass through it.

```py
class CliffordBackend:
    def __init__(self, ..., engine, operations, ...):
        self.engine = Engine
        self.clifford_operations = operations  # not `engine.clifford_operations`

    def _clifford_pre_execution_reshape(self, ...):
        return self.operations._clifford_pre_execution_reshape(...)
        # not
        # ```
        # return self.engine._clifford_pre_execution_reshape(...)
        # ```
```
and if you want to avoid passing both `engine` and `operations` in the `__init__`, then your options are:
- define `ENGINE` in `operations` and use
    ```py
    self.engine = operations.ENGINE
    ```
    (currently is the opposite)
- or define suitable tuples `(engine, operations)` such that the user can just pass the pair defined somewhere in Qibo as a constant (or use the conventional name of that pair)

Reply:
By definition, it's not for the used to choose the operators. The user chooses the `engine` (aka one of the other backends) and the `engine` deals with the rest. And the whole point is to have the `CliffordBackend` to be compatible with all the other engines (aka backends) available in `qibo`. The clifford operations being an input to the user is a no-no.

Reply:
I agree, the operations should not be an input the user has to specify. From my perspective the user should pass the backend they want to use as engine to the `CliffordBackend`, that's it. Therefore, the only other alternative, I guess, is to check which backend was passed as engine and import the correct operations accordingly, which I honestly don't like...
I don't know, to me, the original solutions seems to be the more natural one.

Reply:
> By definition, it's not for the used to choose the operators. The user chooses the `engine` (aka one of the other backends) and the `engine` deals with the rest. And the whole point is to have the `CliffordBackend` to be compatible with all the other engines (aka backends) available in `qibo`. The clifford operations being an input to the user is a no-no.

I'd say that *all the other engines* is not a great design choice, because:
- `CliffordBackend` is a backend itself, and it should not be in any way special (trivially, requesting that is just compatible with all the engines would make it recursive)
- there are (and will be) backends that are not sensible (or simple) for `CliffordBackend` to use as engine, e.g. the TensorFlow backend (unless there is a sensible way to differentiate it), or the tensor networks ones (since they will be tailored for Circuits, not for generic array operations)

I would treat the Clifford engines as a thing on their own. The fact that internally you will reuse backends is an implementation detail, and indeed you are not reusing the backend gates (which are reimplemented by the operations) but just some array operations.

In particular, from what I can see you're using only `.cast()`, and `.np`
https://github.com/qiboteam/qibo/blob/b3e839ef9c8433341999d6ad6bdc068aa07fbdf6/src/qibo/backends/clifford.py#L61
https://github.com/qiboteam/qibo/blob/b3e839ef9c8433341999d6ad6bdc068aa07fbdf6/src/qibo/backends/clifford.py#L260
https://github.com/qiboteam/qibo/blob/b3e839ef9c8433341999d6ad6bdc068aa07fbdf6/src/qibo/backends/clifford.py#L64
(and the second should be `self.cast()`, otherwise is worthless defining it in the first place...)
(and the third should be `self.np`, for the same reason)

EDIT: I forgot to inspect the `Clifford` in `quantum_info` as well, there is also a `self.identity_density_matrix`
https://github.com/qiboteam/qibo/blob/b3e839ef9c8433341999d6ad6bdc068aa07fbdf6/src/qibo/quantum_info/clifford.py#L411-L413

Reply:
I'm not saying you should not use an engine (at least not now), nor that you should change everything in this PR.

But if this PR arrives to the point in which it is unavoidable to modify the engine for Clifford, then I'd suggest taking the time to minimally rearrange the structure, and even just ask the user for an `engine` string (`"numpy"`, `"numba"`, `"cupy"`) irrespectively of the fact that they are also available as Qibo backends (keep this internal).